### ISSUE TITLE
Set application icon

### ISSUE THREAD

**Issue by woodyjon** [therecipe/qt#573]
You say [here](https://github.com/therecipe/qt/wiki/Setting-the-Application-Icon) that application icon can be set by calling SetIcon() on the main window.
But in many examples, you use:
```
gui.NewQGuiApplication(len(os.Args), os.Args)
engine := qml.NewQQmlApplicationEngine(nil)
engine.Load(core.NewQUrl3("qrc:/qml/main.qml", 0))
gui.QGuiApplication_Exec()
```
which is what I use in my app. That way, I don't get how we access the main window to set the icon?
Could you please help me on that point?

Thank you very much!

**Comment by therecipe** [comment]
Hey

Something like this should work:

```go
app := gui.NewQGuiApplication(len(os.Args), os.Args)
app.SetWindowIcon(gui.NewQIcon5("qrc:/qml/icon.ico"))
engine := qml.NewQQmlApplicationEngine(nil)
engine.Load(core.NewQUrl3("qrc:/qml/main.qml", 0))
gui.QGuiApplication_Exec()
```

**Comment by woodyjon** [comment]
Thank you.
Indeed, this builds, contrary to other tests that I had done which did not build.
On linux, although it builds, it does not change the icon in the taskbar (actually it does not seem to do anything). But I will do other tests on windows and mac.

**Comment by therecipe** [comment]
Ah, sorry yes this will only work on windows.
For linux, and mac please take a look here: https://github.com/therecipe/qt/wiki/Setting-the-Application-Icon#setting-the-icon-on-linux and https://github.com/therecipe/qt/wiki/Setting-the-Application-Icon#setting-the-icon-on-macos

(And to set the application icon on windows you will also have to do this: https://github.com/therecipe/qt/wiki/Setting-the-Application-Icon#setting-the-icon-on-windows)

**Comment by woodyjon** [comment]
Thank you!

May I ask a small question, while you're at it? I did not find in the wiki or issues:
In the go code, how can I specify differences depending on the platform to build for? Like when I build for linux, use this path, and for mac, use that other path?

Thank you very much for all this work!!

**Comment by therecipe** [comment]
Yeah, no problem :)

You can let go exclude/include `*.go` files by using build tags: https://golang.org/pkg/go/build/#hdr-Build_Constraints

For example to build the file only on linux you can add `// +build linux` at the top of the file
or to exclude it on linux: `// +build !linux` (above the package statement)

And/or you could also suffix your file names with `_linux.go`, `_windows.go`, ... to archive the same effect.
(I did this here: https://github.com/therecipe/qt/blob/master/qt_android.go for example)

Btw, I just saw you created a GUI for the TRTL coin.
I created some showcase GUI for the SiaCoin to show how to structure QML/Go code properly, but haven't found time to publish it yet. If you are interested, I will try to dig it out and release it later or tomorrow.

**Comment by woodyjon** [comment]
Yes, that's is the project for which I use your work! So thank you very much!
Linux version working. Now I'm on the mac, then the windows one. Main difference I encountered so far is the way mac handles directories. Different from linux.
Yes, I'll definitely look at your siacoin app if you publish it.

For the platform dependent stuff, I was not thinking about full files, but some variables which would have different values depending on the platform. But I guess if I can exclude files, I can just put some variables in those files, and give them different values there. I'll try that now. Thanks!!

**Comment by therecipe** [comment]
Sorry, yes to get the platform info at runtime just use: `runtime.GOOS` or `core.QSysInfo_ProductType()` http://doc.qt.io/qt-5/qsysinfo.html#productType

**Comment by woodyjon** [comment]
Thanks again!

**Comment by therecipe** [comment]
@woodyjon
You can find the sia wallet here: https://github.com/therecipe/qt/tree/master/internal/examples/showcases/sia
But they changed some of their API since I worked on this, so I'm not exactly sure if it is fully usable at the moment.
However, the models are populated with demo data by default so you can checkout most stuff without really running a siad instance.
So the only thing you need to run is `go get github.com/NebulousLabs/Sia/cmd/...` first.

The way this structured is rather complicated on first glance, but most QML Types are implemented in the same structure:
`QML Type (implemented from a Go Template subclassing QQuickItem with qtmoc) <-> Go Controller <-> Go Model`

And only the "Controllers" from the different instances communicate with each other.

The idea behind this structuring is, to be make everything modular by encapsulating the QML Types in a (Qml - Go Template - Go Controller - Go Model) component.
This way it's also possible to switch out the whole Qml (View) part of the component against Widgets if needed, without changing the underlying logic. (And it's also possible to share components across different code bases.)

This concept adds a bit of boilerplate code, mostly connecting the various signals/slots.
But once you internalized this, you can be much more productive. Because you will always know where to find everything.

Also the empty "_" imports in `main.go` are necessary to let go know that all these folders need to be checked for go or qml code, as some of them don't even contain go code but only QML files + modules and are therefor not referenced from go code at all.

However, I will probably create a new page in the wiki to explain everything in more detail someday.

**Comment by woodyjon** [comment]
Ok, great, I'll check that. Lot of stuff to learn in that app then. I am now very motivated to refactor my app and apply your architecture. Hope to find the time. Thank you very much!!

Can I ask another question? When I am developing, I understood I should run `qtdeploy test desktop`, but then I don't have any log or list of bugs. For go, it is alright as I can log in a file (but I don't see the logs for crashes,...). For QML, same, I don't see the bugs. What I do is that if my app builds but does not launch, it means I made a mistake in my QML file, and by trial and errors, I find my mistake. It is not very efficient, I guess there must be a better way. How do you do?

**Comment by therecipe** [comment]
You could install msys2 and then use the mingw shells to start your application, IIRC that will give you access to the std outputs. Or maybe develop on linux/macOS.

Btw some other thing you also can do to speedup the development, is to let your application directly load the qml files instead of the "qrc" files. That way you will only need to restart your application instead of re-compiling it each time after you made some changes to your qml files.
Take a look here: https://github.com/therecipe/qt/blob/master/internal/examples/showcases/sia/main.go#L80
and you will probably need to use this as well: https://github.com/therecipe/qt/blob/master/internal/examples/showcases/sia/main.go#L42
otherwise the qmlengine will create cache files and then don't re-read you qml files after you made changes to them.

**Comment by therecipe** [comment]
Ah, yes the "test" command will not block the terminal to show you the output.
(I will maybe create another command to do this in the future.)

You will need to run something like `qtdeploy build desktop && ./deploy/linux/*.sh` instead.

**Comment by woodyjon** [comment]
Yes, makes sense. Thank you very much for your time assisting noobs ;-)

**Comment by therecipe** [comment]
No problem ;)
